### PR TITLE
Added fault location pathes

### PR DESCRIPTION
**Pull request by gcsantos-gpa**
The method of determining the handedness of line segments isn't great, but I am not sure there are better alternatives. This might require and additional update to the system center line segment wizards.

### CODE REVIEW

File: Source/Libraries/FaultData/DataResources/FaultDataResource.cs [hunk 1]
```diff
@@ -1023,14 +1023,80 @@ private void PopulateFaultInfo(Fault fault, DataGroup dataGroup, VICycleDataGrou
...
+                        bool useBackward = false;
+                        if (forwardPath.TransmissionPath.Segments.Count == 0) continue;
+
+                        // If we have exactly one segment then a backwards path won't be neccessary
+                        if (forwardPath.TransmissionPath.Segments.Count != 1)
+                        {
+                            bool forwardlyChiral= meterTable.QueryRecordsWhere(testQuery, meter.ID, forwardPath.TransmissionPath.Segments[0].ID).Any();
+                            bool backwardlyChiral = meterTable.QueryRecordsWhere(testQuery, meter.ID, forwardPath.TransmissionPath.Segments[forwardPath.TransmissionPath.Segments.Count - 1].ID).Any();
+                            useForward = forwardlyChiral || !backwardlyChiral;
+                            useBackward = !forwardlyChiral || backwardlyChiral;
```
Comment (StephenCWills) [inline Source/Libraries/FaultData/DataResources/FaultDataResource.cs:1056]: Are we envisioning some kind of looped transmission line where the meter is somehow monitoring from both ends of the path simultaneously? It shouldn't be physically possible for a single set of currents to represent current flowing from two physically distinct line segments. Seems to me like `useForward && useBackward` should represent an error, and the path should only be traversed in one direction.

File: Source/Libraries/FaultData/DataResources/FaultDataResource.cs [hunk 2]
```diff
@@ -1023,14 +1023,80 @@ private void PopulateFaultInfo(Fault fault, DataGroup dataGroup, VICycleDataGrou
...
+                                distanceRemaining -= segment.Length;
+                            }
+                            faultPaths.Add(forwardPath);
+                        }
+                        if (useBackward)
+                        {
+                            Fault.Path backwardPath = forwardPath.ShallowCopy();
+                            backwardPath.IsMain = (0 == index) && !useForward;
+                            double distanceRemaining = backwardPath.FaultDistance;
+                            for (int pathIndex = backwardPath.TransmissionPath.Segments.Count - 1; pathIndex >= 0; pathIndex--)
```
Comment (StephenCWills) [inline Source/Libraries/FaultData/DataResources/FaultDataResource.cs:1080]: Consider using `foreach (LineSegment segment in backwardPath.TransmissionPath.Segments.Reverse())`.

File: Source/Libraries/FaultData/DataResources/FaultDataResource.cs [hunk 3]
```diff
@@ -1023,14 +1023,80 @@ private void PopulateFaultInfo(Fault fault, DataGroup dataGroup, VICycleDataGrou
...
                     summary.DistanceAlgorithmIndex = i;
                     summary.DistanceAlgorithm = fault.Curves[i].Algorithm;
                     summary.Distance = fault.Curves[i][calculationCycle].Value;
-                    summary.IsValid = IsValid(summary.Distance, dataGroup);
+                    Line line = Line.DetailedLine(dataGroup.Asset);
+                    List<Fault.Path> faultPaths = new List<Fault.Path>();
+                    for (int index = 0; index < line.Path.Count; index++)
+                    {
+                        // We can break early if this condition is found because any segments after are shorter
+                        if (summary.Distance > line.Path[index].Length) break;
```
Comment (StephenCWills) [inline Source/Libraries/FaultData/DataResources/FaultDataResource.cs:1039]: This doesn't seem to be true. Paths appear to be arbitrarily ordered on account of the list having been generated by enumerating a `Dictionary` without any explicit sorting.

https://github.com/GridProtectionAlliance/openXDA/blob/c5377ffa511e6fec38f851823f0ae596f7a42dbf/Source/Libraries/openXDA.Model/TransmissionElements/Line.cs#L194-L203

**Review by StephenCWills — CHANGES_REQUESTED** [review]
It doesn't look like this PR changes anything about how the fault distance is computed. I notice that you requested Christoph's review so perhaps I'm just out of the loop, but I thought we were going to be computing fault distance by stepping through each segment on each path to adjust the impedance values used for the distance algorithms.

File: Source/Libraries/FaultData/DataResources/FaultDataResource.cs (hunk 3 above)
Comment (gcsantos-gpa) [reply Source/Libraries/FaultData/DataResources/FaultDataResource.cs:1039]: See line 74, it is ordered.

File: Source/Libraries/FaultData/DataResources/FaultDataResource.cs (hunk 1 above)
Comment (gcsantos-gpa) [reply Source/Libraries/FaultData/DataResources/FaultDataResource.cs:1056]: What I am trying to accomplish is to log every possible line segment the fault could be on. Therefore, if there is uncertainty on which direction we need to go, we need to log both.

**Comment by gcsantos-gpa** [comment]
> It doesn't look like this PR changes anything about how the fault distance is computed. I notice that you requested Christoph's review so perhaps I'm just out of the loop, but I thought we were going to be computing fault distance by stepping through each segment on each path to adjust the impedance values used for the distance algorithms.

Looking through I think you may have a good point. It shouldn't be super difficult to do as you describe.

File: Source/Libraries/FaultData/DataResources/FaultDataResource.cs (hunk 2 above)
Comment (gcsantos-gpa) [reply Source/Libraries/FaultData/DataResources/FaultDataResource.cs:1080]: I didn't know this was a thing, thank you for the lesson!

**Review by clackner-gpa — CHANGES_REQUESTED** [review]
As discussed:
(1) Add `Pathnumber` to `FautlCurve` Table
(2) Change `Faultsummary` Table to reference `Linesegment` instead of `Asset`

**Review by StephenCWills — CHANGES_REQUESTED** [review]
I have commits incoming to address these comments.

**Review by StephenCWills — APPROVED** [review]

**Pull request merged**